azure-iot-device/tests/iothub/test_sync_clients.py

```diff
@@ -200,37 +236,201 @@ def test_receive_method_request_default_mode(self, mocker, client, method_name):
         assert inbox_mock.get.call_count == 1
         assert inbox_mock.get.call_args == mocker.call(block=True, timeout=None)
 
-    def test_send_method_response_calls_transport(self, client, pipeline):
+    # TODO: add missing tests
+
+
+class SharedClientSendMethodResponseTests(object):
+    @pytest.mark.it("Begins a 'send_method_response' pipeline operation")
+    def test_send_method_response_calls_pipeline(self, client, pipeline):
         response = MethodResponse(request_id="1", status=200, payload={"key": "value"})
         client.send_method_response(response)
         assert pipeline.send_method_response.call_count == 1
         assert pipeline.send_method_response.call_args[0][0] is response
 
 
-@pytest.mark.describe("IoTHubModuleClient (Synchronous)")
-class TestIoTHubModuleClient(ClientSharedTests):
+################
+# DEVICE TESTS #
+################
+class IoTHubDeviceClientTestsConfig(object):
+    client_class = IoTHubDeviceClient
+
+    @pytest.fixture
+    def client(self, pipeline):
+        return IoTHubDeviceClient(pipeline)
```

Comment:
This class exists to prevent having to add fixtures to all the Device test classes. I'm looking for a better way to do it, since I'd rather avoid multiple inheritance in the test classes, but this works for now at least.

Reply:
Also the config classes are only used for fixtures right ? its only being done to pass client around right ? If the fixtures are defined at top with one being device_client and the other being module client and then used around then probably the Test classes wont need to inherit from the config classes

---
In reply to: [290048211](https://github.com/Azure/azure-iot-sdk-python-preview/pull/101#discussion_r290048211) [](ancestors = 290048211)

Reply:
The problem is that the tests are shared, and thus have to import a fixture with the same name.

For instance, both device and module have "send_event", so there is a SharedClientSendEventTests class. But for the tests to be generic, they need to use a singular client fixture. I can't say "device_client" or "module_client", because the test itself is agnostic.

I could potentially instead have three fixtures at module level, device_client, module_client, and client. client would be a parametrized fixture that returns both device and module client. Then shared tests could rely on client, and module/device specific tests could rely on the specific fixture.

I think that would change how the output is formatted though - I'll need to double check.

---
In reply to: [290088275](https://github.com/Azure/azure-iot-sdk-python-preview/pull/101#discussion_r290088275) [](ancestors = 290088275,290048211)